### PR TITLE
Add FS module support for CustomTarget, CustomTargetIndex, and BuildTarget

### PR DESCRIPTION
**Pull request by tristan957**

### CODE REVIEW

**Comment by tristan957** [comment]
Please let me know if you think other APIs should have their support expanded.

**Comment by tristan957** [comment]
cc @anarazel, not sure if this helps

**Comment by anarazel** [comment]
> cc @anarazel, not sure if this helps

@tristan957 I think it could - right now one of the uses in postgres has separate loops dealing with custom/build targets , various files and just strings. That'd be a easier to do if fs worked uniformly.

However, I don't think the feature really works at the moment?  Just passing a `CustomTarget, CustomTargetIndex, BuildTarget` to str(), results in `fs.name()` returning bogus results:
```
Message: t.full_path(): /srv/dev/build/postgres/m-dev-assert/src/include/catalog/postgres.bki , fs.name(): include', '--set-version=17', '--output=@OUTDIR@', '@INPUT@']>[postgres.bki]>
```

**Comment by tristan957** [comment]
Let me go ahead and add some tests to validate my work. Thanks for the feedback.

**Comment by anarazel** [comment]
@tristan957 The reason for the behaviour I observed is that while `fs.relative_to()` has support to turn `CustomTargetIndex` et al to a path, `fs.name` etc don't.

**Comment by tristan957** [comment]
Hmmm. Must've missed something when I was reviewing my work. I had this shelved for a bit and then came back to it recently.

**Comment by tristan957** [comment]
I never actually implemented the changes, only added the FeatureNews, ....

**Comment by tristan957** [comment]
Added the support. I wonder if we should also add support for generators?

File: mesonbuild/modules/fs.py [hunk 1]
```diff
@@ -189,39 +189,64 @@ def is_samepath(self, state: 'ModuleState', args: T.Tuple['FileOrString', 'FileO
...
+    def stem(self, state: 'ModuleState', args: T.Tuple[T.Union[FileOrString, BuildTargetTypes]], kwargs: T.Dict[str, T.Any]) -> str:
+        arg = args[0]
+        if isinstance(arg, str):
+            original = PurePath(arg)
+        elif isinstance(arg, File):
+            FeatureNew('fs.stem(file)', '0.59.0').use(state.subproject, location=state.current_node)
+            original = PurePath(str(arg))
+        else:
+            FeatureNew('fs.stem(build_tgt | custom_tgt | custom_idx)', '1.3.0').use(state.subproject, location=state.current_node)
+            original = PurePath(state.backend.get_target_filename(arg))
```
Comment (anarazel) [inline mesonbuild/modules/fs.py:249]: This seems all quite redundant - perhaps worth creating new types to deduplicate typed arguments, and moving the type specific handling into a common function that gets the function name as an argument?

**Comment by anarazel** [comment]
> Added the support.

Thanks, seems to work based on some fairly quick testing.

>  I wonder if we should also add support for generators?

Would be nice, the degree to which generator results behave differently from other things is painful...

File: test cases/common/220 fs module/meson.build [hunk 2]
```diff
@@ -5,6 +5,8 @@ is_windows = build_machine.system() == 'windows'
 fs = import('fs')
 
 f = files('meson.build')
+btgt = executable('btgt', 'btgt.c')
+ctgt = fs.copyfile('ctgt.txt')
```
Comment (anarazel) [inline test cases/common/220 fs module/meson.build:9]: Perhaps worth testing the behaviour of a custom target with multiple outputs.

**Comment by tristan957** [comment]
> Would be nice, the degree to which generator results behave differently from other things is painful...

Let me look at generator...

File: mesonbuild/modules/fs.py [hunk 3]
```diff
@@ -108,7 +108,7 @@ def expanduser(self, state: 'ModuleState', args: T.Tuple[str], kwargs: T.Dict[st
     @typed_pos_args('fs.is_absolute', (str, File))
     def is_absolute(self, state: 'ModuleState', args: T.Tuple['FileOrString'], kwargs: T.Dict[str, T.Any]) -> bool:
         if isinstance(args[0], File):
-            FeatureNew('fs.is_absolute_file', '0.59.0').use(state.subproject, location=state.current_node)
+            FeatureNew('fs.is_absolute(file)', '0.59.0').use(state.subproject, location=state.current_node)
```
Comment (eli-schwartz) [inline mesonbuild/modules/fs.py:111]: This should really be `fs.is_absolute with File arg` if you ask me...

File: mesonbuild/modules/fs.py [hunk 4]
```diff
@@ -117,7 +117,7 @@ def is_absolute(self, state: 'ModuleState', args: T.Tuple['FileOrString'], kwarg
     def as_posix(self, state: 'ModuleState', args: T.Tuple[str], kwargs: T.Dict[str, T.Any]) -> str:
         """
         this function assumes you are passing a Windows path, even if on a Unix-like system
-        and so ALL '\' are turned to '/', even if you meant to escape a character
+        and so ALL '\\' are turned to '/', even if you meant to escape a character
```
Comment (eli-schwartz) [inline mesonbuild/modules/fs.py:120]: So the issue here is that we actually do mean `\` and that is what people reading the file see, but pydoc sees  the single-quote being escaped?

```
Help on function as_posix in module mesonbuild.modules.fs:

as_posix(self, state: "'ModuleState'", args: 'T.Tuple[str]', kwargs: 'T.Dict[str, T.Any]') -> 'str'
    this function assumes you are passing a Windows path, even if on a Unix-like system
    and so ALL '' are turned to '/', even if you meant to escape a character
```

IMO the correct solution here is to change the docstring to `f"""`

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

**Comment by tristan957** [comment]
Fucking git

**Comment by tristan957** [comment]
> > Would be nice, the degree to which generator results behave differently from other things is painful...
> 
> Let me look at generator...

I don't think this is possible unless you can index GeneratedList, and I couldn't find any documentation that says you can.

**Pull request merged**